domains/pallets/executive/src/lib.rs

```diff
@@ -402,55 +410,128 @@ where
                 used_weight,
                 DispatchClass::Mandatory,
             );
         }
 
-        <AllPalletsWithSystem as OnFinalize<BlockNumberFor<System>>>::on_finalize(block_number);
+        <AllPalletsWithSystem as OnFinalize<BlockNumberFor<ExecutiveConfig>>>::on_finalize(
+            block_number,
+        );
     }
 
     /// Wrapped `frame_executive::Executive::apply_extrinsic`.
     ///
     /// Note the storage root in the end.
-    pub fn apply_extrinsic(uxt: Block::Extrinsic) -> ApplyExtrinsicResult {
+    pub fn apply_extrinsic(uxt: ExtrinsicOf<ExecutiveConfig>) -> ApplyExtrinsicResult {
         Pallet::<ExecutiveConfig>::push_root(Self::storage_root());
-        let res = frame_executive::Executive::<
-            System,
-            Block,
-            Context,
-            UnsignedValidator,
-            AllPalletsWithSystem,
-            COnRuntimeUpgrade,
-        >::apply_extrinsic(uxt);
+
+        // apply the extrinsic within another transaction so that changes can be reverted.
+        let res = with_storage_layer(|| {
+            frame_executive::Executive::<
+                ExecutiveConfig,
+                BlockOf<ExecutiveConfig>,
+                Context,
+                UnsignedValidator,
+                AllPalletsWithSystem,
+                COnRuntimeUpgrade,
+            >::apply_extrinsic(uxt.clone())
+            .map_err(|err| DispatchError::Other(err.into()))
+        });
+
+        // apply extrinsic failed with transaction validity error
+        // this could happen for following scenarios
+        // - Bad extrinsic Signature
+        //      This extrinsic will be ignored by the operators during the bundle check
+        //      and marks such bundle as Invalid.
+        // - Extrinsic execution failed
+        //      There are multiple scenarios why this can happen
+        //      - Inherent extrinsic failed. If this happens, we should fail to apply extrinsic
+        //      - Pre and Post dispatch fails. Check the test `test_domain_block_builder_include_ext_with_failed_predispatch`
+        //        why this could happen. If it fail due to this, then we revert the inner storage changes
+        //        but still include extrinsic so that we can clear inconsistency between block body and trace roots.
+        let res = match res {
+            Ok(dispatch_outcome) => Ok(dispatch_outcome),
+            Err(err) => {
+                let encoded = uxt.encode();
+                let (maybe_signer, dispatch_info) =
+                    ExecutiveConfig::ExtrinsicStorageFees::extract_signer(uxt);
+                // if this is not a normal extrinsic, then transaction should not execute
+                // we should fail here.
+                if dispatch_info.class != DispatchClass::Normal {
+                    return Err(TransactionValidityError::Invalid(
+                        InvalidTransaction::BadMandatory,
+                    ));
+                }
```

Comment:
Besides `Normal` and `Mandatory` there is also `Operational` which is possible to fail